### ISSUE TITLE
Consider saving workflow when program starts

### ISSUE THREAD

**Issue by bruno-f-cruz** [bonsai-rx/bonsai#2077]
It is my understanding that the editor can keep a cached state of the workflow that has not yet been serialized to disk. In cases where asserting the state of the software is important, e.g. asserting if the repository is dirty at start-up, this feature makes it difficult to know what is the state that run the experiment as the cached state is not serialized to disk. One way to go around this may be to "force" the workflow to save itself each time the program starts.

**Comment by glopesdev** [comment]
Feedback from DCM:
- Temp files instead of saving on run? Would this help?
- Have a toggle button for "save on run"?
  - Save in global settings
- It is possible to develop "buggy" custom extensions which will build, save, but not deserialize correctly from XML after save, is this a problem?

**Comment by PathogenDavid** [comment]
> * Temp files instead of saving on run? Would this help?

I don't think this is especially helpful for Bruno's original stated use-case when this issue was made. IIRC he was embedding the hash of the workflow in logs so there was traceability as to which workflow was used for the experiment.

It might make sense for Bonsai to do this in the background as a method for auto-recovery though.

> * Have a toggle button for "save on run"? 

If this is being envisioned as a button on the toolbar this sounds like extreme clutter to me. The behavior Bruno is describing is exactly what Visual Studio and VSCode do.

IIRC Visual Studio *used* to prompt you before saving (with a checkbox to remember the choice) but these days it just saves automatically.

Not opposed to the behavior being optional, but Bonsai really needs a proper settings menu for this sort of thing.

> * It is possible to develop "buggy" custom extensions which will build, save, but not deserialize correctly from XML after save, is this a problem?

IMO it's a bad thing that it's even possible to run a workflow that won't be able to be deserialized. It honestly might not even be a bad thing for compilation to be done on a copy of the workflow that has gone through a serialize→deserialize trip. (This is probably a separate issue to consider though.)

**Comment by jonnew** [comment]
Related to #https://github.com/open-ephys/bonsai-onix1/issues/385.

FWIW I think having this functionality in one way or another is very important for an experiment reproducibility standpoint and support standpoint so that its known what code and parameters produced a particular set of data files.

**Comment by glopesdev** [comment]
Feedback from DCM:
  - Could be useful for several other issues
  - Consider some kind of comparison with existing file and save only if there are differences

**Comment by glopesdev** [comment]
For a consistent experience, it would be important to resolve #2248 first, so that the workflow is not capable of self-modifying persistent dynamic state to disk. This is a confusing and error-prone experience, especially in cases where the workflow is expected to only be executed from a baseline and never deliberately saved.

**Comment by jonnew** [comment]
Hi All, I just want to comment that I think this is an important issue for our users.  Some of our users are asking for the ability to save e.g. a JSON file per bonsai run with some hardware metadata. This is redundant and less powerful compared to saving experimental metadata in .bonsai file is the correct solution since its a nearly complete experiment description.

**Comment by glopesdev** [comment]
@jonnew While this issue is valid on its own right, we've been moving away further and further from having the workflow do double-duty as both program and configuration. This is becoming more important as workflows grow more general and data-driven themselves. The reality is that the .bonsai file can never grow to encompass all possible configuration data, which will necessarily need to include binary files, data queried from databases and other sources at runtime, etc.

Especially for per-session data, given we are now version controlling workflows next to source code in experiments, we really don't want to have diffs in history which are purely per-session configuration, so in general we are not anymore relying on this feature to replace the ability to import or export hardware configuration data.
